### PR TITLE
Migrate inspector ui to null safety.

### PR DESCRIPTION
**Pull request by polina-c**

### CODE REVIEW

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 1]
```diff
@@ -57,12 +55,12 @@ const overflowEpsilon = 0.1;
 ///    sum(renderSize) = maxSizeAvailable
 ///
 List<double> computeRenderSizes({
-  @required Iterable<double> sizes,
-  @required double smallestSize,
-  @required double largestSize,
-  @required double smallestRenderSize,
-  @required double largestRenderSize,
-  @required double maxSizeAvailable,
+  required Iterable<double?> sizes,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:58]: can this be non-nullable?

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 2]
```diff
@@ -143,69 +140,83 @@ class LayoutProperties {
...
-  final String description;
-  final num flexFactor;
-  final FlexFit flexFit;
-  final bool isFlex;
-  final Size size;
+  final BoxConstraints? constraints;
+  final String? description;
+  final num? flexFactor;
+  final FlexFit? flexFit;
+  final bool? isFlex;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:157]: make non-nullable

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 3]
```diff
@@ -143,69 +140,83 @@ class LayoutProperties {
...
+  final Size? size;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:158]: can size be made non-nullable? that would eliminate several null checks and asserts elsewhere

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 4]
```diff
@@ -275,8 +286,8 @@ extension LayoutPropertiesExtension on LayoutProperties {
 
 final Expando<FlexLayoutProperties> _flexLayoutExpando = Expando();
 
-extension MainAxisAlignmentExtension on MainAxisAlignment {
-  MainAxisAlignment get reversed {
+extension MainAxisAlignmentExtension on MainAxisAlignment? {
+  MainAxisAlignment? get reversed {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:290]: non nullable on both lines

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 5]
```diff
@@ -329,29 +340,28 @@ class FlexLayoutProperties extends LayoutProperties {
   }) : super(node);
 
   factory FlexLayoutProperties.fromDiagnostics(RemoteDiagnosticsNode node) {
-    if (node == null) return null;
     // Cache the properties on an expando so that local tweaks to
     // FlexLayoutProperties persist across multiple lookups from an
     // RemoteDiagnosticsNode.
-    return _flexLayoutExpando[node] ??= _buildNode(node);
+    return (_flexLayoutExpando[node] ??= _buildNode(node))!;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:346]: what if _buildNode(node) returns null

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 6]
```diff
@@ -429,8 +444,8 @@ class FlexLayoutProperties extends LayoutProperties {
 
   String get type => direction.flexType;
 
-  num get totalFlex {
-    if (children?.isEmpty ?? true) return 0;
+  num? get totalFlex {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:447]: return int instead of num

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 7]
```diff
@@ -439,62 +454,62 @@ class FlexLayoutProperties extends LayoutProperties {
   }
 
   Axis get crossAxisDirection {
-    return direction == Axis.horizontal ? Axis.vertical : Axis.horizontal;
+    return direction;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:457]: revert this change. We need to return the opposite direction for crossAxisDirection

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 8]
```diff
@@ -439,62 +454,62 @@ class FlexLayoutProperties extends LayoutProperties {
...
   }
 
-  double get mainAxisDimension => dimension(direction);
+  double? get mainAxisDimension => dimension(direction);
 
-  double get crossAxisDimension => dimension(crossAxisDirection);
+  double? get crossAxisDimension => dimension(crossAxisDirection);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:462]: these can be made non-nullable if `size` above can be made non-nullable

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 9]
```diff
@@ -143,69 +140,83 @@ class LayoutProperties {
...
-    return displayChildren?.map((child) => child.dimension(axis))?.toList();
+  List<double?> childrenDimensions(Axis axis) {
+    return displayChildren.map((child) => child.dimension(axis)).toList();
   }
 
-  List<double> get childrenWidths => childrenDimensions(Axis.horizontal);
+  List<double?>? get childrenWidths => childrenDimensions(Axis.horizontal);
 
-  List<double> get childrenHeights => childrenDimensions(Axis.vertical);
+  List<double?>? get childrenHeights => childrenDimensions(Axis.vertical);
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:185]: childrenDimensions returns a non-nullable list, so these can too

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 10]
```diff
@@ -143,69 +140,83 @@ class LayoutProperties {
...
 
-  double get height => size?.height;
+  double? get height => size?.height;
 
-  double dimension(Axis axis) => axis == Axis.horizontal ? width : height;
+  double? dimension(Axis axis) => axis == Axis.horizontal ? width : height;
 
-  List<double> childrenDimensions(Axis axis) {
-    return displayChildren?.map((child) => child.dimension(axis))?.toList();
+  List<double?> childrenDimensions(Axis axis) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:179]: non-nullable once size is non-nullable

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 11]
```diff
@@ -143,69 +140,83 @@ class LayoutProperties {
...
+  bool get hasChildren => children.isNotEmpty;
 
-  double get width => size?.width;
+  double? get width => size?.width;
 
-  double get height => size?.height;
+  double? get height => size?.height;
 
-  double dimension(Axis axis) => axis == Axis.horizontal ? width : height;
+  double? dimension(Axis axis) => axis == Axis.horizontal ? width : height;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:177]: non-nullable once size is non-nullable

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 12]
```diff
@@ -439,62 +454,62 @@ class FlexLayoutProperties extends LayoutProperties {
...
+    return height! + overflowEpsilon < max(childrenHeights!.cast<double>());
   }
 
   bool get startIsTopLeft {
-    assert(direction != null);
     switch (direction) {
       case Axis.horizontal:
         switch (textDirection) {
           case TextDirection.ltr:
+          case null:
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:485]: this shouldn't be necessary since direction is non-nullable

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 13]
```diff
@@ -439,62 +454,62 @@ class FlexLayoutProperties extends LayoutProperties {
...
             return true;
           case TextDirection.rtl:
             return false;
         }
-        break;
       case Axis.vertical:
         switch (verticalDirection) {
           case VerticalDirection.down:
+          case null:
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:493]: we should remove this and try to make verticalDirection non-nullable.

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 14]
```diff
@@ -439,62 +454,62 @@ class FlexLayoutProperties extends LayoutProperties {
...
   List<RenderProperties> childrenRenderProperties({
-    @required double smallestRenderWidth,
-    @required double largestRenderWidth,
-    @required double smallestRenderHeight,
-    @required double largestRenderHeight,
-    @required double Function(Axis) maxSizeAvailable,
+    required double? smallestRenderWidth,
+    required double? largestRenderWidth,
+    required double? smallestRenderHeight,
+    required double? largestRenderHeight,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:507]: I think these can all be non-nullable

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 15]
```diff
@@ -439,62 +454,62 @@ class FlexLayoutProperties extends LayoutProperties {
...
+    required double Function(Axis) maxSizeAvailable,
   }) {
     /// calculate the render empty spaces
-    final freeSpace = dimension(direction) - sum(childrenDimensions(direction));
+    final freeSpace = dimension(direction)! -
+        sum(childrenDimensions(direction).cast<double>());
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:512]: this cast shouldn't be necessary once childrenDimensions is changed to return `List<double>`

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 16]
```diff
@@ -556,8 +571,8 @@ class FlexLayoutProperties extends LayoutProperties {
         /// include free space in the computation
         sizes.add(freeSpace);
       }
-      final smallestSize = min(sizes);
-      final largestSize = max(sizes);
+      final smallestSize = min(sizes.cast<double>());
+      final largestSize = max(sizes.cast<double>());
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:575]: same

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 17]
```diff
@@ -680,28 +695,31 @@ class FlexLayoutProperties extends LayoutProperties {
   }
 
   List<RenderProperties> crossAxisSpaces({
-    @required List<RenderProperties> childrenRenderProperties,
-    @required double Function(Axis) maxSizeAvailable,
+    required List<RenderProperties>? childrenRenderProperties,
+    required double Function(Axis)? maxSizeAvailable,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:699]: non nullable

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 18]
```diff
@@ -680,28 +695,31 @@ class FlexLayoutProperties extends LayoutProperties {
...
       space.crossAxisDimension =
           maxSizeAvailable(crossAxisDirection) - space.crossAxisDimension;
       if (space.crossAxisDimension <= 0.0) continue;
       if (crossAxisAlignment == CrossAxisAlignment.center) {
         space.crossAxisDimension *= 0.5;
-        space.crossAxisRealDimension *= 0.5;
+        final crossAxisRealDimension = space.crossAxisRealDimension;
+        if (crossAxisRealDimension != null) {
+          space.crossAxisRealDimension = crossAxisRealDimension * 0.5;
+        }
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:722]: this can be reverted once height and width are made non nullable

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 19]
```diff
@@ -747,20 +765,20 @@ class RenderProperties {
         dx = offset?.dx ?? 0.0,
         dy = offset?.dy ?? 0.0;
 
-  final Axis axis;
+  final Axis? axis;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:768]: non nullable

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 20]
```diff
@@ -747,20 +765,20 @@ class RenderProperties {
...
 
   /// represents which node is rendered for this object.
-  LayoutProperties layoutProperties;
+  LayoutProperties? layoutProperties;
 
   double dx, dy;
   double width, height;
-  double realWidth, realHeight;
+  double? realWidth, realHeight;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:775]: non nullable

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart [hunk 21]
```diff
@@ -804,21 +822,21 @@ class RenderProperties {
...
       realHeight = newVal;
     }
   }
 
-  double get crossAxisRealDimension =>
+  double? get crossAxisRealDimension =>
       axis == Axis.horizontal ? realHeight : realWidth;
 
-  set crossAxisRealDimension(double newVal) {
+  set crossAxisRealDimension(double? newVal) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:839]: non nullable

**Review by kenzieschmoll — COMMENTED** [review]
initial review on inspector_data_models.dart. Making changes for those review comments will likely affect the other files in this PR, so I'll wait to review the rest until the comments for inspector_data_models are addressed.

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 1 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:58]: Nope. It gets down to LayoutProperties's width and height that are nullable by design.

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 2 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:157]: done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 3 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:158]: Done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 4 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart]: done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 5 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:345]: moved null-check to _buildNode, where it seems to be better fit

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 6 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:445]: in one of descendants it can return double:
https://github.com/flutter/devtools/blob/f7e30d461ac79d0090bfe780e4f5399ee8978097/packages/devtools_app/lib/src/screens/inspector/layout_explorer/flex/utils.dart#L138

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 7 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:455]: done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 8 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:460]: done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 9 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart]: done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 10 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart]: done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 11 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart]: done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 12 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:483]: direction is not nullable, but textDirection is nullable.

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 14 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:505]: done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 16 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:573]: done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 18 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:720]: done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 19 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:766]: done

File: packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart (hunk 21 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/inspector_data_models.dart:837]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart [hunk 22]
```diff
@@ -286,33 +285,33 @@ class AnimatedLayoutProperties<T extends LayoutProperties>
...
-      _lerpList(begin.childrenWidths!, end.childrenWidths!);
+  List<double> get childrenWidths =>
+      _lerpList(begin.childrenWidths, end.childrenWidths).cast<double>();
 
   @override
   BoxConstraints? get constraints {
     try {
       return BoxConstraints.lerp(
-          begin.constraints, end.constraints, animation!.value);
+          begin.constraints, end.constraints, animation.value);
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:314]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart [hunk 23]
```diff
@@ -40,33 +40,35 @@ abstract class LayoutExplorerWidgetState<W extends LayoutExplorerWidget,
...
 
-  AnimationController entranceController;
-  CurvedAnimation entranceCurve;
-  AnimationController changeController;
+  late AnimationController entranceController;
+  late CurvedAnimation entranceCurve;
+  late AnimationController changeController;
 
-  CurvedAnimation changeAnimation;
+  CurvedAnimation? changeAnimation;
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:47]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart [hunk 24]
```diff
@@ -40,33 +40,35 @@ abstract class LayoutExplorerWidgetState<W extends LayoutExplorerWidget,
...
-  RemoteDiagnosticsNode get selectedNode =>
-      inspectorController?.selectedNode?.value?.diagnostic;
+  RemoteDiagnosticsNode? get selectedNode =>
+      inspectorController.selectedNode?.value?.diagnostic;
 
   InspectorController get inspectorController => widget.inspectorController;
 
-  InspectorService get inspectorService => serviceManager.inspectorService;
+  InspectorService? get inspectorService =>
+      serviceManager.inspectorService as InspectorService?;
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:67]: nope. 
serviceManager.inspectorService is InspectorServiceBase?, that is supertype for Inspector Service?

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart [hunk 25]
```diff
@@ -86,48 +88,49 @@ abstract class LayoutExplorerWidgetState<W extends LayoutExplorerWidget,
...
 
   /// Whether this layout explorer can work with this kind of node.
-  bool shouldDisplay(RemoteDiagnosticsNode node);
+  bool shouldDisplay(RemoteDiagnosticsNode? node);
 
-  Size get size => properties.size;
+  Size? get size => properties!.size;
 
-  List<LayoutProperties> get children => properties.displayChildren;
+  List<LayoutProperties>? get children => properties!.displayChildren;
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:95]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart [hunk 26]
```diff
@@ -86,48 +88,49 @@ abstract class LayoutExplorerWidgetState<W extends LayoutExplorerWidget,
   }
 
   /// Whether this layout explorer can work with this kind of node.
-  bool shouldDisplay(RemoteDiagnosticsNode node);
+  bool shouldDisplay(RemoteDiagnosticsNode? node);
 
-  Size get size => properties.size;
+  Size? get size => properties!.size;
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:93]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart [hunk 27]
```diff
@@ -86,48 +88,49 @@ abstract class LayoutExplorerWidgetState<W extends LayoutExplorerWidget,
...
-      assert(objectGroupManager.next == nextObjectGroup);
-      objectGroupManager.promoteNext();
+      assert(manager.next == nextObjectGroup);
+      manager.promoteNext();
     }
     return computeLayoutProperties(node);
   }
 
-  L computeLayoutProperties(RemoteDiagnosticsNode node);
+  L computeLayoutProperties(RemoteDiagnosticsNode? node);
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:120]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart (hunk 22 above)
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart [hunk 28]
```diff
@@ -275,14 +278,14 @@ abstract class LayoutExplorerWidgetState<W extends LayoutExplorerWidget,
...
-  Future<Object> onForceRefresh() async {
+  Future<Object?> onForceRefresh() async {
     _setProperties(await fetchLayoutProperties());
     return null;
   }
 
   /// Currently this is not working so we should listen to controller selection event instead.
   @override
-  Future<void> onInspectorSelectionChanged() {
+  Future<void>? onInspectorSelectionChanged() {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:288]: both of these should just have the return type `Future<void>` and not add a return statement

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart [hunk 29]
```diff
@@ -106,14 +104,15 @@ class BorderLayout extends StatelessWidget {
 
 @immutable
 class Truncateable extends StatelessWidget {
-  const Truncateable({Key key, this.truncate, this.child}) : super(key: key);
+  const Truncateable({Key? key, this.truncate = false, this.child})
+      : super(key: key);
 
-  final Widget child;
+  final Widget? child;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:110]: child should be non null. Look at all uses of the `Truncateable` widget. null is never passed in for `child`, so we know it should be non-nullable. Also, we are unconditionally null-asserting in the build method, which is another code smell that something should probably be non-nullable.

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart [hunk 30]
```diff
@@ -125,25 +124,24 @@ class Truncateable extends StatelessWidget {
...
-  })  : assert(title != null),
-        super(key: key);
+  }) : super(key: key);
 
-  final LayoutProperties layoutProperties;
+  final LayoutProperties? layoutProperties;
   final String title;
-  final Widget child;
-  final Widget hint;
+  final Widget? child;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:139]: child should be required and non-null. Same comment as above. Look at references for WidgetVisualizer and look which values are passed in. All non-nulalble.

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart [hunk 31]
```diff
@@ -125,25 +124,24 @@ class Truncateable extends StatelessWidget {
...
+    required this.layoutProperties,
     this.child,
     this.overflowSide,
     this.largeTitle = false,
-  })  : assert(title != null),
-        super(key: key);
+  }) : super(key: key);
 
-  final LayoutProperties layoutProperties;
+  final LayoutProperties? layoutProperties;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:137]: non-nullable. I don't see anywhere that null is being passed in for layoutProperties.

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart [hunk 32]
```diff
@@ -284,52 +282,65 @@ class AnimatedLayoutProperties<T extends LayoutProperties>
     return _children;
   }
 
-  List<double> _lerpList(List<double> l1, List<double> l2) {
+  List<double?> _lerpList(List<double?> l1, List<double?> l2) {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:285]: Look at the references for this method. I don't see anywhere that we pass in a List of type `double?`.

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart [hunk 33]
```diff
@@ -362,13 +373,15 @@ class AnimatedLayoutProperties<T extends LayoutProperties>
   double get height => size.height;
 
   @override
-  bool get isFlex => begin.isFlex && end.isFlex;
+  bool get isFlex => begin.isFlex == true && end.isFlex == true;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:376]: isFlex is non nullable, so we do not need to use `== true`

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/widget_constraints.dart [hunk 34]
```diff
@@ -17,40 +15,41 @@ import 'utils.dart';
...
-    @required this.child,
+    required this.properties,
+    double? arrowHeadSize,
+    required this.child,
     this.warnIfUnconstrained = true,
   }) : arrowHeadSize = arrowHeadSize ?? defaultIconSize;
 
-  final Widget child;
-  final LayoutProperties properties;
+  final Widget? child;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/widget_constraints.dart:24]: make child non-nullable and use a SizedBox instead where null is passed in currently

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/widget_constraints.dart [hunk 35]
```diff
@@ -17,40 +15,41 @@ import 'utils.dart';
...
+  final LayoutProperties? properties;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/widget_constraints.dart:25]: I think these can be non-null as well

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart (hunk 32 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:285]: this should definitely be non-null. In general, having `List<int?>`, List<bool?>` or List<double?>` is a code smell and should probably be a lint.

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart [hunk 36]
```diff
@@ -86,48 +88,49 @@ abstract class LayoutExplorerWidgetState<W extends LayoutExplorerWidget,
...
+      manager.promoteNext();
     }
     return computeLayoutProperties(node);
   }
 
   L computeLayoutProperties(RemoteDiagnosticsNode node);
 
   AnimatedLayoutProperties<L> computeAnimatedProperties(L nextProperties);
-  void updateHighlighted(L newProperties);
+  void updateHighlighted(L? newProperties);
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:123]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart [hunk 37]
```diff
@@ -185,7 +188,7 @@ abstract class LayoutExplorerWidgetState<W extends LayoutExplorerWidget,
   }
 
   // update selected widget and trigger selection listener event to change focus.
-  void refreshSelection(RemoteDiagnosticsNode node) {
+  void refreshSelection(RemoteDiagnosticsNode? node) {
```
[the remark this answers is not in the document]
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:191]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart (hunk 28 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/layout_explorer_widget.dart:288]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart (hunk 29 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:110]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart (hunk 30 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:139]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart (hunk 32 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:285]: In this case yes. In general, I do not see why such code smells. Some elements may have value not defined. For example, in a column, some elements have height specified and others need to share what is remaining equally.
Am I missing something?

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart (hunk 31 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:137]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart (hunk 33 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart:376]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/widget_constraints.dart (hunk 34 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/widget_constraints.dart:24]: done

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/widget_constraints.dart (hunk 35 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/widget_constraints.dart:25]: done

**Review by kenzieschmoll — APPROVED** [review]

**Comment by polina-c** [comment]
tested

File: packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart (hunk 32 above)
Comment (jacob314) [reply packages/devtools_app/lib/src/screens/inspector/layout_explorer/ui/utils.dart]: Maybe my comment was too extreme. There are a few cases where it can be handy to have a nullable list of a primitive type but in general I think it is a code smell in the context of migrating code. Any time I see a `List<double?>` I'd want a strong reason why it has to be nullable. This is because there are natural sentinel values that often make sense to use (e.g. `0` or `NaN`) rather than null so you really have to have a good reason to make the values nullable.

Searching for uses of List<bool?> and List<int?> across code bases, I see a lot more cases where it indicates the null safety migration could have been done better than cases where List<int?> seems like a reasonable design call. Often the list doesn't really need to be nullable as all content is expected to eventually have a non-null value or there was just a bug in running the migration tool so List<int?> snuck in.


For example, if you have a `List<bool?>` and really need null values, probably defining an enum with 3 values would be cleaner. For the Column case, is using `0` or another marker value for omitted height constraints cleaner?
